#773: Add steps to install Rancher Extensions in an air-gapped environment (#807)
rancher/rancher-docs: docs/integrations-in-rancher/rancher-extensions.md

:::info

> If you've upgraded from an older Rancher version to v2.7.0, uninstalled built-in extensions are no longer listed under the **Available** tab, unless you've manually imported them.

Reviewer: ```suggestion
In v2.7.0, the built-in extensions will not be displayed under the **Available** tab. Therefore, you will need to manually add the desired repos to install extensions.
```

The original text is to indicate a limitation found in certain versions (2.7.0-2.7.?) where users would expect something that's built-in to be available for selection. The update here changes the meaning to indicate that there's a change in existing behavior. Also, extensions were introduced in 2.7 so there is no upgrade scenario involved.

    ![Manage repositories](/img/manage-repos.png)

5. Under the **Available** tab, click **Install** on the desired extension and version as in the example below. You can also update your extension from this screen, as the button to **Update** will appear on the extension if one is available.

    ![Install Kubewarden](/img/install-kubewarden.png)

6. Click the **Reload** page button that will appear after your extension successfully installs. Note that a logged-in user who has just installed an extension will not see a change to the UI **unless** they reload the page.

    ![Reload button](/img/reload-button.png)

> ### Importing and Installing Extensions in an Air-Gapped Environment

Reviewer 2: Nit: Importing and Installing Extensions Repository in an Air-Gapped Environment

> 1. Find the address of the container image you want to import as an extension. Rancher provides some extensions, such as Kubewarden and Elemental, through the `ui-plugin-catalog` container image at https://hub.docker.com/r/rancher/ui-plugin-catalog/tags. You should import and use the latest tagged version of the image to ensure you receive the latest features and security updates.

Reviewer 3: Find the address of the container image you want to import as an extension _repository_.

> 1. Find the address of the container image you want to import as an extension. Rancher provides some extensions, such as Kubewarden and Elemental, through the `ui-plugin-catalog` container image at https://hub.docker.com/r/rancher/ui-plugin-catalog/tags. You should import and use the latest tagged version of the image to ensure you receive the latest features and security updates.

Author: ```suggestion
1. Find the address of the container image repository  that you want to import as an extension. Rancher provides some extensions, such as Kubewarden and Elemental, through the `ui-plugin-catalog` container image at https://hub.docker.com/r/rancher/ui-plugin-catalog/tags. You should import and use the latest tagged version of the image to ensure you receive the latest features and security updates.
```

>     * **(Optional)** If the container image is private: [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-UI-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.

Reviewer 3: The namespace **must** be all lower case, i.e. `cattle-ui-plugin-system`.

>     * **(Optional)** If the container image is private: [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-UI-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.

Author: ```suggestion
    * **(Optional)** If the container image is private: [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-ui-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.
```

1. Click **☰**, then select **Extensions**, under **Configuration**.

1. On the top right, click **⋮ > Manage Extension Catalogs**.

1. Select the **Import Extension Catalog** button.

1. Enter the image address in the **Catalog Image Reference** field. 

    * **(Optional)** If the container image is private: Select the secret you just created from the **Pull Secrets** drop-down menu.

1. Click **Load**. The extension will now be **Pending**.

1. Return to the **Extensions** page.

1. Select the **Available** tab, and click the **Reload** button to make sure that the list of extensions is up to date.

1. Find the extension you just added, and click the **Install** button.

> ## Uninstalling Extensions

Reviewer 2: Nit: Uninstalling Extensions Repositories

    ![Reload button](/img/reload-button.png)

### Importing and Installing Extensions in an Air-Gapped Environment

> 1. Find the address of the container image you want to import as an extension.

Reviewer 2: Rancher provides some extensions such as Kubewarden and elemental and we need to tell the user about this. The extensions are available in this container image https://hub.docker.com/r/rancher/ui-plugin-catalog/tags

The latest available tag is recommended to be used and imported.

> 1. [Create](../how-to-guides/new-user-guides/kubernetes-resources-setup/secrets.md) a registry secret within the `cattle-UI-plugin-system` namespace. Enter the domain of the image address in the **Registry Domain Name** field.

Reviewer 2: There is not always a secret needed. For example, for publicly available image such as https://hub.docker.com/r/rancher/ui-plugin-catalog/tags, there is no need to secret to be created. Can we mention this somehow ?

:::

## Updating and Upgrading Extensions

1. Click **☰ > Extensions** under **Configuration**.
1. Select the **Updates** tab. 
1. Click **Update**.

If there is a new version of the extension, there will also be an **Update** button visible on the associated card for the extension in the **Available** tab.

> ### Updating and Upgrading Extensions in an Air-gapped Environment

Reviewer 2: Updating and Upgrading Extensions Repository

> ### Updating and Upgrading Extensions in an Air-gapped Environment

Author: ```suggestion
### Updating and Upgrading an Extensions Repository in an Air-gapped Environment
```

1. Click **☰**, then click on the name of your local cluster.
1. From the sidebar, select **Workloads > Deployments**.
1. From the namespaces dropdown menu, select **All Namespaces**.
1. Find the **cattle-ui-plugin-system** namespace. 

> <!-- TK Waiting for answer -->

Reviewer 2: Find the deployment 
ui-plugin-catalog

Click on edit config 
change rancher/ui-plugin-catalog:1.0.1 to rancher/ui-plugin-catalog:1.0.2 (1.0.1 and 1.0.2 are just examples)

click on save.

Reviewer 2: Also, they need to mirror the image rancher/ui-plugin-catalog:1.0.2 to their private registry before dpoing this.

:::

## Deleting Helm Charts

1. Click **☰**, then click on the name of your local cluster.
1. From the sidebar, select **Apps > Installed Apps**.
1. Find the name of the chart you want to delete and select the checkbox next to it. 

> 1. Click **Delete**.

Reviewer 3: Why is this section here? You shouldn't need to delete an extension chart from the Installed Apps page.

## Updating and Upgrading Extensions

1. Click **☰ > Extensions** under **Configuration**.
1. Select the **Updates** tab. 
1. Click **Update**.

If there is a new version of the extension, there will also be an **Update** button visible on the associated card for the extension in the **Available** tab.

### Updating and Upgrading an Extensions Repository in an Air-gapped Environment

> Extensions repository that aren't air-gapped are automatically updated. If the repository is air-gapped, you must update it manually.

Reviewer 3: _Extension repositories_ that aren't air-gapped...

> Extensions repository that aren't air-gapped are automatically updated. If the repository is air-gapped, you must update it manually.

Author: ```suggestion
Extensions repositories that aren't air-gapped are automatically updated. If the repository is air-gapped, you must update it manually.
```

> First, mirror the latest changes to your private repository by following the same steps for initially [importing and installing an extension repository](#importing-and-installing-extensions-in-an-air-gapped-environment).

Reviewer 3: ...changes to your private _registry_ by following...

> First, mirror the latest changes to your private repository by following the same steps for initially [importing and installing an extension repository](#importing-and-installing-extensions-in-an-air-gapped-environment).

Author: ```suggestion
First, mirror the latest changes to your private registry by following the same steps for initially [importing and installing an extension repository](#importing-and-installing-extensions-in-an-air-gapped-environment).
```

After you mirror the latest changes, follow these steps:

1. Click **☰ > Local**.
1. From the sidebar, select **Workloads > Deployments**.

> 1. From the namespaces dropdown menu, select **All Namespaces**.

Reviewer 3: It is unnecessary to select All Namespaces, you should just select the `cattle-ui-plugin-system` namespace.

> 1. From the namespaces dropdown menu, select **All Namespaces**.

Author: ```suggestion
1. From the namespaces dropdown menu, select **cattle-ui-plugin-system**.
```

1. Find the **cattle-ui-plugin-system** namespace. 
1. Select the `ui-plugin-catalog` deployment.
1. Click **⋮ > Edit config**.

> 1. Update the **Container image** field with the container of the deployment.

Reviewer 3: This is slightly confusing, it should read: "Update the Container Image field within the deployment's container with the latest image."

> 1. Update the **Container image** field with the container of the deployment.

Author: ```suggestion
1. Update the **Container Image** field within the deployment's container with the latest image.
```

1. Click **Save**.

## Deleting Extension Repositories

1. Click **☰ > Extensions** under **Configuration**.
1. On the top right, click **⋮ > Manage Repositories**.

> 1. Find the name of the extension repository you want to delete. Select the checkbox next to the repository name, then click **Delete**.

Reviewer 3: This section should provide instructions on deleting the Extension Catalog, not the repository. Deleting an imported extension catalog will automatically remove the associated repository, deployment, and service.

> 1. Find the name of the extension repository you want to delete. Select the checkbox next to the repository name, then click **Delete**.

Reviewer 3: Is this related to Air-gapped extension repositories that are created when importing an extension catalog?
